resources/type-system/inference.md

```diff
@@ -315,45 +320,60 @@ described below with a typing context as computed above.
+    of `e`, using the local type inference algorithm described below with typing
+    context `K`, and update `T` to be `UP(S, T)`.
+  - If the enclosing function is marekd `async`, for each `return e;` statement
+    in the block, let `S` be the inferred type of `e`, using the local type
+    inference algorithm described below with typing context `FutureOr<K>`,
+    and update `T` to be `UP(flatten(S), T)`.
+  - If the enclosing function is a non-generator function, for each `return;`
+    statement in the block, update `T` to be `UP(Null, T)`.
   - For each `yield e;` statement in the block, let `S` be the inferred type of
     `e`, using the local type inference algorithm described below with typing
     context `K`, and update `T` to be `UP(S, T)`.
   - If the enclosing function is marked `sync*`, then for each `yield* e;`
     statement in the block, let `S` be the inferred type of `e`, using the
     local type inference algorithm described below with a typing context of
-    `Iterable<K>`; let `E` be the type such that `Iterable<E>` is a
-    super-interface of `S`; and update `T` to be `UP(E, T)`.
+    `Iterable<K>`. If there exists a type `E` such that `Iterable<E>` is a
+    super-interface of `S`, update `T` to be `UP(E, T)`. Otherwise update
+    `T` to be `UP(S, T)`.
+    _It is a compile-time error if *S* is not a assignable to
+    `Iterable<Object?>`, so either *S* implements `Iterable`, or it is one of
+    `dynamic` or `Never`._
   - If the enclosing function is marked `async*`, then for each `yield* e;`
     statement in the block, let `S` be the inferred type of `e`, using the
     local type inference algorithm described below with a typing context of
-    `Stream<K>`; let `E` be the type such that `Stream<E>` is a super-interface
-    of `S`; and update `T` to be `UP(E, T)`.
+    `Stream<K>`. If there exists a type `E` such that `Stream<E>` is a
+    super-interface of `S`, update `T` to be `UP(E, T)`. Otherwise update
+    `T` to be `UP(S, T)`.
+    _It is a compile-time error if *S* is not a assignable to
+    `Stream<Object?>`, so either *S* implements `Iterable`, or it is one of
+    `dynamic` or `Never`._
 
 The **actual returned type** of the function literal is the value of `T` after
 all `return` and `yield` statements in the block body have been considered.
 
 Let `T` be the **actual returned type** of a function literal as computed above.
 Let `R` be the greatest closure of the typing context `K` as computed above.
 
-With null safety: if `R` is `void`, or the function literal is marked `async`
-and `R` is `FutureOr<void>`, let `S` be `void` (without null-safety: no special
-treatment is applicable to `void`).
+With null safety, if `R` is `void`, let `S` be `void`
+_(without null-safety: no special treatment is applicable to `void`)_.
 
-Otherwise, if `T <: R` then let `S` be `T`.  Otherwise, let `S` be `R`.  The
+Otherwise (_without null safety or if `R` is not `void`_),
```

Comment:
Then we haven't specified `S` if `R` is void and without null safety.

We have one clause:
If (with null safety && R is void): S = `void`
And an otherwise, which should cover both of:
* with null safety && R is not void
* without null safety
(aka "without null safety or if R is not void")